### DISCUSSION TITLE
How does highlighting work?

### DISCUSSION THREAD

**Discussion by anuvyklack** [nvim-neorg/neorg#345] [Q&A] [answered]
How do highlighting groups are set?                                             
                                                                                
The pending item highlight properties are somehow set to yellow and italic:     
            
![Screenshot-20220131143726-536x134](https://user-images.githubusercontent.com/13056013/151851643-57c4ba5f-9e67-4017-a5b0-0114dcdc4043.png)
                                              
                                                                                
The TSplaygroud shows the next                                                   

![Screenshot-20220131143758-1282x258](https://user-images.githubusercontent.com/13056013/151851790-6f0673ee-87f2-4e81-b78b-d819d698cdd9.png)

my concealer module settings are:                                               
                                                                                
```lua                                                                          
pending = {                                                                     
   enabled = true,                                                              
   icon = "",                                                                  
   highlight = "NeorgTodoItemPendingMark",                                      
   query = "(todo_item_pending) @icon",                                         
   extract = function() return 1 end,                                           
},                                                                              
```                                                                             
                                                                                
But I cannot find any mentions about `NeorgTodoItemPendingMark` or `NeorgTodoItemPending` or `NeorgTodoItem1Pending` inside my colorsheme [gruvbox-material](https://github.com/sainnhe/gruvbox-material) code base or inside Neorg code base, except `queries/norg/highlights.scm` file.  How dose Neovim deside which highlight properties to choose? 
                                                                                
The second question is: what `extract` function do?                             
```lua                                                                          
  extract = function() return 1 end,                                            
```

**Comment by vhyrro** [comment, marked as answer]
Hey! Before I answer the question about the highlight groups I do want to let you know that you only need to change the settings that you care about in the `pending` table - everything else will just be kept as the default. This means that you can refactor your code to:
```lua
pending = {                                                                                        
   icon = "",
},  
```
And not have to worry about all the magic like `extract`. That function basically tells Neorg how many characters it should offset by, in this case one, otherwise you would get: `<icon>-]` instead of `[<icon>]`.

About the highlights though - Neorg uses `core.integrations.treesitter` to bind Neorg highlight groups to treesitter highlight groups to feel consistent across all colourschemes that support it. To view every highlight enter a .norg file and run `:filter Neorg highlight` - this'll show you all the hl groups and what Neorg binds them to. To view a single highlight you can run `:hi NeorgTodoItem1Pending`, which for me shows that it binds to `TSNamespace`.

These can be changed in two ways. The first is to simply go somewhere in your config and
```lua
	vim.cmd([[
		hi link NeorgTodoItem1Pending MyPreferredHighlightGroup
	]])
```
Neorg will recognize that you've set the highlight group manually and won't touch it.

You can also customize `core.integrations.treesitter` directly, like so:
```lua
["core.integrations.treesitter"] = {
	config = {
		highlights = {
			-- all highlights go here
		}
	}
}
```

To view the entire table of highlight groups, consult [the wiki entry](https://github.com/nvim-neorg/neorg/wiki/Treesitter-Integration#configuration). It's currently not indented properly because of some limitations but that'll show you how highlight groups can be changed through the config table.

I hope that answers your question, but when in doubt do ask again! I'll try and help out whenever I can :)
